### PR TITLE
Add reverse csc and test script

### PR DESCRIPTION
**Pull request by GalaxySciTech**
Add reverse csc and test script

### CODE REVIEW

File: contracts/libraries/ReverseHeaderReader.sol [hunk 1]
```diff
@@ -0,0 +1,698 @@
...
+            addresses[i] = addr;
+            dataIndex += 20;
+        }
+
+        return addresses;
+    }
+
+    /*
+     * @param rlp-encoded block header.
+     * @return (currentValidator list, nextValidator list).
```
Comment (wgr523) [inline contracts/libraries/ReverseHeaderReader.sol]: return doesn't have "currentValidator list"

File: contracts/ReverseFullCheckpoint.sol [hunk 2]
```diff
@@ -0,0 +1,427 @@
+// SPDX-License-Identifier: MIT
+pragma solidity =0.8.19;
+
+import {ReverseHeaderReader as HeaderReader} from "./libraries/ReverseHeaderReader.sol";
+
+contract ReverseFullCheckpoint {
+    // Compressed subnet header information stored on chain
```
Comment (wgr523) [inline contracts/ReverseFullCheckpoint.sol]: I thought reverse checkpoint is about "mainnet header" not "subnet header"

File: contracts/ReverseFullCheckpoint.sol [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+            stateRoot: v2esbnBlock.stateRoot,
+            transactionsRoot: v2esbnBlock.transactionsRoot,
+            parentHash: v2esbnBlock.parentHash,
+            mix: (uint256(v2esbnBlock.number) << 128) |
+                (uint256(v2esbnBlock.roundNumber) << 64) |
+                uint256(block.number)
+        });
+        validators[v2esbn] = Validators({
+            set: next,
+            threshold: int256((next.length * 2 * 100) / 3)
```
Comment (wgr523) [inline contracts/ReverseFullCheckpoint.sol]: @liam-lai please help check threshold here

File: contracts/ReverseFullCheckpoint.sol [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+            );
+            require(
+                headerTree[validationParams.parentHash].mix != 0,
+                "Parent Missing"
+            );
+            require(
+                int256(
+                    uint256(
+                        uint64(
+                            headerTree[validationParams.parentHash].mix >> 128
```
Comment (wgr523) [inline contracts/ReverseFullCheckpoint.sol]: I have a question. there are many `headerTree[validationParams.parentHash]`. If we define a variable to store it, will it reduce gas? Like this
```
x = headerTree[validationParams.parentHash]
require(x.mix!=0)
require(...)
```

File: contracts/ReverseFullCheckpoint.sol [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+                next.length > 0 &&
+                validationParams.prevRoundNumber <
+                validationParams.roundNumber -
+                    (validationParams.roundNumber % INIT_EPOCH)
+            ) {
+                setLookup(next);
+
+                validators[validationParams.number] = Validators({
+                    set: next,
+                    threshold: int256((next.length * 2 * 100) / 3)
```
Comment (wgr523) [inline contracts/ReverseFullCheckpoint.sol]: here threshold is 2/3. @liam-lai Please check that whether threshold must be 2/3 in golang code, thx

Comment (wgr523) [reply contracts/ReverseFullCheckpoint.sol]: According to https://github.com/XinFinOrg/XDPoSChain/blob/master/params/config.go#L44-L46
the threshold can be changed. not necessarily *2/3.
In golang code, it is a float, such as 0.667

Comment (GalaxySciTech) [reply contracts/ReverseFullCheckpoint.sol]: fixed

Comment (wgr523) [reply contracts/ReverseFullCheckpoint.sol]: the threshold can be changed.
for example, the community decide to change threshold to 0.500 in golang code. So i suggest in smart contract, make threshold changeable too.

Comment (wgr523) [reply contracts/ReverseFullCheckpoint.sol]: take [https://github.com/XinFinOrg/XDPoSChain/blob/master/params/config.go](https://github.com/XinFinOrg/XDPoSChain/blob/master/params/config.go) as an example, see below code. The threshold change from 0.667 (default) to 0.4 (7956000 round number) to 0.667 (7974000  round number). I'm afraid in mainnet the threshold is changed in the future.
```
DevnetV2Configs = map[uint64]*V2Config{
		Default: {
			MaxMasternodes:       108,
			SwitchRound:          0,
			CertThreshold:        0.667,
			TimeoutSyncThreshold: 3,
			TimeoutPeriod:        30,
			MinePeriod:           2,
		},
		7956000: { // 2024.01.17 Devnet Deplyment Issue
			MaxMasternodes:       108,
			SwitchRound:          7956000,
			CertThreshold:        0.4,
			TimeoutSyncThreshold: 3,
			TimeoutPeriod:        30,
			MinePeriod:           2,
		},
		7974000: {
			MaxMasternodes:       108,
			SwitchRound:          7974000,
			CertThreshold:        0.667,
			TimeoutSyncThreshold: 3,
			TimeoutPeriod:        30,
			MinePeriod:           2,
		},
	}
```

Comment (wgr523) [reply contracts/ReverseFullCheckpoint.sol]: seems the problem can be solved with upgradable proxy. resolve

**Review by wgr523 — APPROVED** [review]

**Pull request merged**